日報の言及機能を作るの課題を提出する

yokomaru opened the pull request:
# 必須要件
- [x] Aさんの日報の本文テキストにBさんの日報のURL（ http://localhost:3000/reports/44 など）が含まれていると、Bさんの日報に「この日報に言及している日報」としてAさんの日報が表示される
    - [x] Qiitaの「この記事は以下の記事からリンクされています」と同じようなイメージ
- [x] 検知すべきドメイン（host）は http://localhost:3000 だけでOK
- [x] 日報の新規作成時だけでなく更新時や削除時も言及が更新される（本文テキストから日報のURLがなくなったら言及もなくなる）

# 歓迎要件
- [x] 日報の本文テキストにURLが含まれていた場合は自動的にリンクにする

# スクショ
- [x] 実装した機能の動きがわかるスクショを載せる
- [x] rubocopとerblintをパスさせる（要スクリーンショット）

https://bootcamp.fjord.jp/products/20316

yokomaru committed:
first commit

yokomaru committed:
add:  gem rinku

日報の本文でリンク表示させるためのgem rinku を追加し、bundle installした

yokomaru committed:
add: create Mention and relations with report

- 言及モデルとテーブルの作成
- 日報とのリレーションを追加

yokomaru committed:
add: save and update with mentions

- 日報内に他の日報へ言及があった場合に追加する処理を追加
    - 日報への言及が更新されたら対応する
 - 更新対象の日報IDを取得するprivateメソッドを追加
- contollerのメソッド変更

yokomaru committed:
add: diaplay mentioned reports

- メンションされた日報の表示を追加
- 現在の時刻と日報の作成日の差分を求め経過時間を表示するメソッドをhelperに追加
- 本文にリンクがある場合はリンクを有効にする処理をhelperに追加
- ユーザー画像の表示(variant追加＋CSS対応)
- メンションがない時はメンションがない旨表示する
- i18n対応

yokomaru committed:
update: display time difference

- 時差を表示するメソッドを経過時間を表示するメソッドという考え方に変えてメソッド名や変数名を変更
- 各分岐の判断文や結果の処理をメソッド化
- 時刻の数値を定数にする
- i18n化

yokomaru committed:
update: error hundling and message with mention methods

- 処理が重複しているためlogger.errorをメソッド化
- エラーメッセージをi18n対応

yokomaru committed:
add: update some reports with mentions

- 言及の初期表示のためいくつかの日報に最初から言及がついたデータを作成する

yokomaru committed:
update: url regexp format

- 比較する回数を減らすために、URI.extractを使わずに正規表現でidを取得する
- 検知するURL正規表現を定数に移動

yokomaru committed:
update: pluralize time unit

- 時間の単位の複数形を対応
    - Qiitaのように時間経過の単位は英語表記のみ
    - 上記をメソッド化
    - 不要になったi18n対応を削除

yokomaru committed:
fix: spell miss mentioned

- mentionnedになっていた箇所をmentionedに変更

yokomaru committed:
update: add unique relation to mention

- メンションにreport_idとmentioned_report_idの組み合わせを一位にするバリデーションを追加

yokomaru committed:
update: display mention created_at in report show

- 日報の詳細で表示している言及一覧で表示している経過時間を日報の作成時間ではなくメンションの作成時間で現在時刻と計算して表示するようにする
    - 該当の日報をメンションしている日報をインスタンス変数で取得
    - 該当の日報をメンションした時間(mentionsテーブルのcreated_at)を取得する処理をモデルに追加

yokomaru commented on app/models/report.rb:
```diff
@@ -3,15 +3,62 @@
   validates :title, presence: true
   validates :content, presence: true
 
+  FORMAT_REPORT_URL = %r{http://localhost:3000/reports/(\d+)/?}
+
```
本文に`http://localhost:3000/reports/(ID)`という文字列があればURLとして認識して処理を行なっておりますが、
以下のようにリンクとして機能しなくなってしまうような表記の場合でも
今回の課題では言及として含めて問題ないでしょうか？

<img width="587" alt="スクリーンショット 2024-07-14 23 05 43" src="https://github.com/user-attachments/assets/4e3e8be9-b4ea-442e-9e89-12270383d312">

```ruby
http://localhost:3000/reports/48http://localhost:3000/reports/48
http://localhost:3000/reports/48です
```

maedana replied:
独断ですけど、この課題の本質的なところからは外れた話なので気にしなくてよしとします。

yokomaru replied:
ご確認ありがとうございます。承知いたしました。

yokomaru committed:
update: add order by mentioned_reports

- 言及一覧のならび方をidの降順に設定

maedana commented on db/schema.rb:
```diff
@@ -59,6 +59,16 @@
 
+  create_table "mentions", force: :cascade do |t|
+    t.integer "report_id"
+    t.integer "mentioned_report_id"
```
外部キーはどちらもnot null制約がいるかなーと思います

yokomaru replied:
こちら以下migrationファイルに` null: false`を追加いたしました。

>20240710142729_create_mentions.rb

maedana commented on app/models/report.rb:
```diff
@@ -3,15 +3,62 @@
+      save!
+      self.mentioning_report_ids = fetch_report_ids(content)
+    end
+  rescue ActiveRecord::RecordInvalid => e
```
このような例外の使い方をするときもありますけど、原則は例外を一種のGoto文のように使うのは避けたいです。今回の課題においては例外に頼らずにメソッドの戻り値で判定して実装してみて欲しいです。`update_with_mentions`についても同様です。

参考
https://qiita.com/jnchito/items/3ef95ea144ed15df3637

yokomaru replied:
参考資料共有ありがとうございます！

`transaction`で囲んだ場合、
エラーが起きた際に`if @report.save`の分岐に引っかからないと勘違いしておりました💦
(そのため、`rescue`で囲って明示的に`false`を返しておりました)
改めて挙動を確認し、エラーが起きた際は`nil`が帰ってくることで`false`の分岐にいくことを確認いたしました。

また、`save`、`update`は`transaction`で囲まなくても
トランザクション内で処理されるとのことだったので、
`update_with_mentions`、`save_with_mentions`自体を削除し、
コールバックの`after_save`で`save`と`update`の際にメンションを保存する処理を追加しました。

maedana commented on app/models/report.rb:
```diff
@@ -3,15 +3,62 @@
+
+  def fetch_report_ids(content)
+    registered_report_ids = Report.all.ids
+    report_ids = content.scan(FORMAT_REPORT_URL).flatten.map(&:to_i).map do |report_id|
```
flatten + mapは https://docs.ruby-lang.org/ja/latest/method/Enumerable/i/collect_concat.html が使えないですかね

maedana replied:
map内部でif分岐する処理はfilter_map使って処理すればcompact削れないですかね。

参考
https://docs.ruby-lang.org/ja/latest/method/Enumerable/i/filter_map.html

yokomaru replied:
メソッドの共有ありがとうございます！

`filter_map`は、おっしゃるとおり`true`になるものだけを配列にしてくれるので`compact`を削ることができました！

`collect_concat`に関しましては、`flaten` + `map`と同じ結果になったのですが、ブロック内に渡されてくる値に違いがありました。

参考：[Ruby の flat\_map は flatten\.map と同じ挙動にはならない \- Just do IT](https://k-koh.hatenablog.com/entry/2022/07/22/090611)

```ruby
content = "http://localhost:3000/reports/52\nhttp://localhost:3000/reports/53\nhttp://localhost:3000/reports/54\nhttp://localhost:3000/reports/55"
# flaten + map
# flatten後のmapブロック内は文字列
(ruby) content.scan(FORMAT_REPORT_URL).flatten.map{|id| p id }
# "52"
# "53"
# "54"
# "55"
# ["52", "53", "54", "55"]

# collect_concat
# collect_concatブロック内は配列が渡される
content.scan(FORMAT_REPORT_URL).collect_concat{|id| p id }
# ["52"]
# ["53"]
# ["54"]
# ["55"]
# ["52", "53", "54", "55"]
```

`collect_concat`を使う場合、ブロック内で配列として扱う必要があったため、
現在の処理をそのまま置き換えようとすると以下のように若干読みづらくなってしまいました。

```ruby
content.collect_concat{|id| id[0].to_i }

```

そのため、`collect_concat`は使わず、`map(&:to_i)`を削除し、`filter_map`のブロック内で`to_i`を行うことで、
二重mapになっていたところを修正いたしました。

```ruby
# 修正後
content.scan(FORMAT_REPORT_URL).flatten.uniq.filter_map do |report_id|
  id = report_id.to_i
  id if registered_report_ids.include?(id)
end
    
# 修正前
report_ids = content.scan(FORMAT_REPORT_URL).flatten.map(&:to_i).map do |report_id|
  report_id if registered_report_ids.include?(report_id)
end
report_ids.compact.uniq

```

maedana commented on app/helpers/reports_helper.rb:
```diff
@@ -0,0 +1,46 @@
+  end
+
+  # 現在の時刻と日報の作成日の差分を求め経過時間を表示
+  def display_passed_time(created_at)
```
自前でやらなくても https://api.rubyonrails.org/v7.1.3.4/classes/ActionView/Helpers/DateHelper.html#method-i-time_ago_in_words を使えばいいのではないかな、と思いました。

yokomaru replied:
このような便利なメソッドがあるのを存じ上げませんでした！今後は実装する前にメソッドを探してみるようにします！

maedana commented on app/models/report.rb:
```diff
@@ -3,15 +3,62 @@
+      report_id if registered_report_ids.include?(report_id)
+    end
+    report_ids.compact.uniq
+  end
```
自分自身に言及出来てしまうようにみえますがいかがでしょうか? もしそうなら自分自身には言及出来ないようにしたいですね。

yokomaru replied:
ご指摘ありがとうございます。
自分自身への言及についてはプラクティスの要件に記載がなかったため考慮できておりませんでした🙏💦

ご指摘いただいた**自分自身**は、保存処理元の日報のみを指しますでしょうか？それとも作成ユーザー(が作成した日報全部)になりますでしょうか？

yokomaru replied:
↑こちら昨日質問させていただいたのですが、自分が作成した別の日報からメンションすることはあり得ると思い直したので、一旦**保存処理元の日報**からのメンションは作成しないように修正いたしました。

maedana replied:
質問に気がづいておらず失礼しました。github上だけですと本業の大量の通知に埋もれがちなのです :cry:

yokomaru committed:
fix: delete display_passed_time method and apply time_ago_in_words method

- 時間差を求めるメソッドがRailsにあったため、自作のメソッドを削除
- i18n対応

yokomaru committed:
fix: add  `null: false` to create mentions migration file

- mentionsの外部キーがnull許容してしまっていたので`null:false`を追加

yokomaru committed:
fix: fetch_report_ids

- mapをfilter_mapに変更
- map(&:to_i)の処理を削除し、filter_map内でto_iを行うようにする
- uniqの位置を修正

yokomaru committed:
fix: delete rescue ActiveRecord::RecordInvalid

- エラーキャッチ部分を削除し、transactionに囲むだけにする
- write_error_logsメソッドを削除する

yokomaru committed:
update: use callback

- transactionで囲んでいたが、saveやupdateは自動的にトランザクション内で処理されるので削除
- saveもupdateも共通の処理になるのでmentions作成の処理をコールバックにする
- 呼び出し元のcontollerも修正

yokomaru committed:
add: check comparing myself

- 自分自身のレポートidと同じidが含まれていた場合はメンションを作成しない

maedana commented on app/models/report.rb:
```diff
@@ -3,15 +3,41 @@
+  private
+
+  def fetch_report_ids(content)
+    registered_report_ids = Report.ids
```
前回指摘しきれてなかったのですが、この実行ですと`Report`のデータが大量にあるとき(数百万件とか)に`registered_report_ids.include?(report_id)`が結構遅いかもしれません。パフォーマンスを考慮した場合は、先にURIからidを集めてきて、それをReportのwhere区に渡してDB上で絞り込むほうが速いのではないかな、と思いました。

なお今回はそこまで求めないのでご対応は任意で結構です。
